src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+        TypeToUniverseDeclaration(datalog::ArgumentType(
+            datalog::ArgumentType::Kind::kNumber, "Number")),
+        TypeToUniverseDeclaration(datalog::ArgumentType(
+            datalog::ArgumentType::Kind::kPrincipal, "Principal")),
+    };
+    ret.merge(builtin_universes);
+    return ret;
+  }
+
+  // This visitor is used only to implmenet GetUniverseDeclarations
+  class UniverseDeclarationGenerationVisitor
+      : public AuthLogicAstTraversingVisitor<
+            UniverseDeclarationGenerationVisitor,
+            absl::flat_hash_set<datalog::RelationDeclaration>> {
+   private:
+    using DeclSet = absl::flat_hash_set<datalog::RelationDeclaration>;
+    DeclSet GetDefaultValue() override { return {}; }
+
+    DeclSet CombineResult(DeclSet left, DeclSet right) override {
+      left.merge(std::move(right));
+      return left;
+    }
+
+    DeclSet Visit(
+        const datalog::RelationDeclaration& relation_declaration) override {
+      DeclSet universe_declarations;
+      for (const auto& argument : relation_declaration.arguments()) {
+        universe_declarations.emplace(
+            TypeToUniverseDeclaration(argument.argument_type()));
+      }
+      return universe_declarations;
+    }
+  };
+
+  //----------------------------------------------------------------------------
+  // Grounding Condition Transformation
+  //----------------------------------------------------------------------------
+  // This class transforms SaysAssertions so that if
+  // there are any potentially ungrounded variables on the LHS of an assertion,
+  // conditions are added to the RHS that check that this variable
+  // is in a universe. As a way of safely overapproximating the ungrounded
+  // variables, this just assumes any *non-constant* arguments in the RHS
+  // are ungrounded, and adds conditions for them to the LHS.
+  // It implements just one public method,
+  // `AddUniverseConditions(SaysAssertion says_assertion)`
+  // that returns a transformed version of the says assertion.
+  // Internally, it extends a visitor for a part of this translation
+
+  class UniverseConditionTransformer
+      : public AuthLogicAstTraversingVisitor<UniverseConditionTransformer,
+                                             std::vector<BaseFact>> {
+   public:
+    UniverseConditionTransformer(const DeclEnv& decl_env)
+        : decl_env_(decl_env) {}
+
+    // The `AddUniverseConditions` methods, including this one,
+    // are not implemented as actual visits because these methods
+    // are implementing a transformation, and the returned values
+    // need to be placed in a part of the syntactic context that
+    // is specific to each syntactic element.
+    SaysAssertion AddUniverseConditions(const SaysAssertion& says_assertion) {
+      std::vector<Assertion> assertion_list;
+      for (const Assertion& assertion : says_assertion.assertions()) {
+        assertion_list.push_back(AddUniverseConditions(assertion));
+      }
```

Comment:
This could maybe be made more concise with `map_iter` from `//src/common/utils`.

Reply:
I think it's about the same to do it with the for loop and would rather leave it as is if that's ok?

Reply:
Sure, that's fine.